### ISSUE TITLE
A/V freeze at Fairplay-encrypted sections with text if clear-lead is not zero

### ISSUE THREAD

**Issue by salmoro** [google/shaka-packager#769]
## System info
**Operating System**: MacOS 10.13.6 / Safari 13.1. Can also repro on IPhone6.
**Shaka Packager Version**: 2.4.2

## Issue and steps to reproduce the problem

**Packager Command**:

```bash
./packager \
'in=./360_600_264.mp4,stream=video,init_segment=./dist/video/360_init.mp4,segment_template=./dist/video/360_$Number$.mp4,hls_name=360,playlist_name=video/360.m3u8,hls_group_id=video' \
'in=./64_eng_264.mp4,stream=audio,init_segment=./dist/audio/eng_init.mp4,segment_template=./dist/audio/eng_$Number$.mp4,language=en,hls_name=eng,playlist_name=audio/eng.m3u8,hls_group_id=audio' \
'in=./en.vtt,stream=text,segment_template=./dist/text/en_$Number$.vtt,language=en,hls_name=en,playlist_name=text/en.m3u8,hls_group_id=text' \
--enable_raw_key_encryption \
--clear_lead 15 \
--keys label=HD:key_id=bf8813b5abaf4265891624b37f2e8bf0:key=9A677F82EAED028162C5B64251FB2758,label=AUDIO:key_id=bf8813b5abaf4265891624b37f2e8bf0:key=9A677F82EAED028162C5B64251FB2758,label=SD:key_id=bf8813b5abaf4265891624b37f2e8bf0:key=9A677F82EAED028162C5B64251FB2758 \
--iv 276BCB2F394C52EB691320F3AC43A3FE \
--protection_scheme cbcs \
--protection_systems FairPlay \
--hls_key_uri 'skd://fps.ezdrm.com/;bf8813b5-abaf-4265-8916-24b37f2e8bf0' \
--hls_master_playlist_output ./dist/master.m3u8
```

**What is the expected result?**
Video should play start to finish without interruption.

**What happens instead?**
Audio and Video freeze at first encrypted section where there is text.
See [here](https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/player/index.html) the video playing in Shaka Player. Notice how at about 20 seconds in (right after clear lead ends) the video freezes but subtitles and currentTime continue progressing and at about 38 seconds (after last subtitle) the video continues playing normally.

Notes:
- Tried on several other Apple devices all of which reproduced this issue.
- Issue is not isolated to Shaka Player. Was also able to reproduce on JW-Player.
- Gaps in subtitles (even 10 seconds) don't resume playback until last subtitle has completed.

Here are the FFMPEG commands I used to create these files:

```bash
wget -O input.mp4 https://images-assets.nasa.gov/video/NHQ_2020_0110_Dreamed%20of%20This/NHQ_2020_0110_Dreamed%20of%20This~orig.mp4
wget -O en.vtt https://temp-for-testing-shlomo.s3.amazonaws.com/en.vtt

./ffmpeg \
-i ./input.mp4 \
-map 0 -dn -an -vf scale=-2:360 -c:v libx264 -level:v 3.0 -profile:v baseline -pix_fmt yuv420p -x264-params scenecut=0:open_gop=0:min-keyint=72:keyint=72 \
-minrate 600k -maxrate 600k -bufsize 600k -b:v 600k -threads 4 -y \
./360_600_264.mp4

./ffmpeg \
-i ./input.mp4 \
-map 0:a:0 -metadata:s:a:0 language=eng -dn -vn -c:a aac -b:a 64k -threads 4 -y \
./64_eng_264.mp4
```

Please LMK if there's anything else I can provide to help you get to the bottom of this.
Thanks!

**Comment by kqyang** [comment]
@salmoro It might be caused by DISCONTINUITY tag mismatch in audio/video and text. Does it work if you remove the clear lead, i.e. `--clear_lead=0`?

**Comment by salmoro** [comment]
@kqyang Yes. Removing clear lead makes the video play well.
Is there any work-around if I want to use clear lead?

**Comment by kqyang** [comment]
It may work by inserting a DISCONTINUITY tag in the beginning of the subtitle playlist.

**Comment by salmoro** [comment]
@kqyang Adding DISCONTINUITY tag in the text playlist seems to fix this. Thank you!

**Comment by kqyang** [comment]
@salmoro Do you insert the DISCONTINUITY tag at the beginning or some point past the clear lead? 

If you insert it in the beginning, does the subtitle work correctly for both clear lead portion and after clear lead?

**Comment by salmoro** [comment]
@kqyang Originally yes I've placed it at the beginning and all seemed well.

Now that you've asked this question I realized that I've only tried that fix on content that the subtitles begin after the clear lead section so I went back and tried the discontinuity fix (at the beginning of the playlist) for the same content referenced above (with subtitles starting before clear-lead end) and noticed that it doesn't completely fix the issue as while the A/V plays well the subtitles don't begin until after the clear lead section.

I've tested moving the DISCONTINUITY tag to the approximate position of 'where the clear lead ends and that seems to solve the above issue.

Here's a diff of the first and second attempt:
![Code_2020-06-03_13-26-39](https://user-images.githubusercontent.com/20197680/83626293-e0e87280-a59d-11ea-8265-a0cf94ad53ad.png)

Here are links to see the player in all three scenarios:

[Original issue](https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/player/index.html): AV freeze at ~20s (a bit after clear lead end).

[Discontinuity fix 1](https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/player/index.html?manifest=https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/discontinuity-fix/master.m3u8): Added DISCONTINUITY tag at beginning of text playlist. Results in AV freeze fix but subtitles only starting at ~18 seconds (right after clear-lead end).

[Discontinuity fix 2](https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/player/index.html?manifest=https://temp-for-testing-shlomo.s3.amazonaws.com/fp-issue/discontinuity-fix2/master.m3u8): Added DISCONTINUITY tag to approximate position of clear-lead end. Results in AV fix and subtitles correctly playing from start (~12 seconds in)

**Comment by salmoro** [comment]
@kqyang 
Yea I guess we can disable "clear lead" and once this is resolved just repackage our content.

From the discussion [here](https://developer.apple.com/forums/thread/113261) it seems that clear-lead can be achieved in HLS but would also require some more work on changes in the playlist. But that in itself would probably not hold us back as we can do some post-processing to the playlists after packaging.

Thank you

**Comment by AwokeKnowing** [comment]
In case it helps anyone, i found codecs were put in random order by shaka so i just have streams named by number and reorder them on command line. And I also had this clearlead issue and here is some code that moves the key to after the first section. this works as long as your clearlead is less than the first segment (eg clearlead 5  for a 6 seconds segments).  if you had longer clearlead, you'd need to move they key down a after a couple segments


`


#put streams in order eg stream_0 - stream_20
m3u8content=$(<video.m3u8)
grep -B 50 -A 1 "stream_0.m3u8" > video.m3u8 <<< $m3u8content
for i in $(seq 1 50); do grep -B 1 "stream_$i.m3u8" >> video.m3u8 <<< $m3u8content; done

#shaka currently doesn't put key in after clear segments with clear lead, so we do it if clearlead
for f in $(ls 'stream'*'.m3u8'); do
    m3u8content=$(<$f)
    grep -B 50 -A 0 '#EXT-X-MAP:URI' > $f <<< $m3u8content
    grep -B 1 -A 1 -m 1 '#EXT-X-BYTERANGE:' >> $f <<< $m3u8content
    grep -B 1 -A 0 'skd://' >> $f <<< $m3u8content
    grep -B 1 -A 100000 -m 1 -E '#EXT-X-BYTERANGE:[0-9]+$' >> $f <<< $m3u8content
done
`

**Comment by kqyang** [comment]
Per the discussion in https://github.com/google/shaka-packager/issues/918#issuecomment-812595437, clear lead is not very well supported with fairplay. Closing this issue as won't fix. If anyone has concerns, please let us know.
